text/0000-custom-model-classes.md

```diff
@@ -0,0 +1,158 @@
+- First step in separating the record implementation from schema information
+
+## Detailed design
+
+After the work in [https://www.notion.so/Errors-RFC-8e9e1fd4e2474ed5a8539ed343047784](https://www.notion.so/Errors-RFC-8e9e1fd4e2474ed5a8539ed343047784),  [https://www.notion.so/Request-State-RFC-b994c09fe3094f2b92e4ff1dd36fa882](https://www.notion.so/Request-State-RFC-b994c09fe3094f2b92e4ff1dd36fa882), [https://www.notion.so/Record-State-on-Record-Data-RFC-3fd05bcd2a83439f96fbcf2bc0f3a581](https://www.notion.so/Record-State-on-Record-Data-RFC-3fd05bcd2a83439f96fbcf2bc0f3a581) we have enough coverage in public apis that we can implement the entire DS.Model class with just a few changes to Store APIs. 
+
+We will need four separate changes: 
+
+- A way to instantiate custom Records and have those records be notified of changes in Record Data
+- A way to access schema information not from the Record Class
+- Replacement of few store methods that take DS.Models
+- Deprecate passing DS.Model classes to adapter/serializers/exposing them on Snapshots
+
+## Instantiating custom Records
+
+First we need to expose a method on the store for instantiating a record, and way for the record to get notified when RecordData values change:
+
+```ts
+interface RecordDataWrapper { // proxies a subset of RD methods and hides the rest
+  performMutation(operation: Operation): void
+      
+  getAttr(identifier, key)
+  isAttrDirty(identifier, key)
+    
+  changedAttributes(identifier)
+  hasChangedAttributes(identifier)
+  rollbackAttributes(identifier)
+  
+  getRelationship(identifier, key)
+  
+  setRecordId(identifier, id)
+  
+  getErrors(identifier)
+  getMeta(identifier)
+  
+  isNew(identifier)
+  isDeleted(identifier)
+}
+  
+interface RecordDataFor {
+  (identifier: RecordIdentifier): RecordDataWrapper
+}
+    
+interface NotificationCallback {
+  (identifier: Identifier, notificationType: 'attributes' | 'relationships' | 'errors'  | 'meta'  | 'unload') => void;
+}
+    
+interface NotificationManager {
+  subscribeAttributeChanges(identifier, NotificationCallback)
+  subscribeRelationshipChanges(identifier, NotificationCallback)
+  subscribeErrorsChanges(identifier, NotificationCallback)
+  subscribeMetaChanges(identifier, NotificationCallback)
+  subscribeUnload(identifier, NotificationCallback)
+  unsubscribe(identifier, NotificationCallback)
+ }
+    
+class Store {
+  instantiateRecord(
+    identifier: RecordIdentifier, 
+    createArgs: CreateArgs, 
+    recordDataFor: RecordDataFor, 
+    notificationManager: NotificationManager): unknown)
+}
+```
+
+Instead of exposing the entire RecordData Class, we expose only the local facing methods and hide the server/adapter facing methods that the Model should not have access to, by giving a RecordData lookup method to `instantiateRecord`
+
+## Exposing schema information
```

Comment:
There is no discussion about dynamic/static expectations for these schemas.  It is somewhat implied that they can be dynamic as the function signatures take identifiers, but we should be explicit about this.

Reply:
Added